### ISSUE TITLE
Does linkskip still work in net7.0-ios?

### ISSUE THREAD

**Issue by dlprows** [xamarin/xamarin-macios#18740]
I created a new project with `dotnet new ios` 

I added to the sln a class library with a class that is just there to serialize/deserialize
I added a similar class with `[Preserve(AllMembers = true)]` to the main ios project that is once again just there for serializing and deserializing

I set the link mode to Full, and added `--linkskip=ClassLibrary1` to the `MtouchExtraArgs`

When i debug, the class with the preserve attribute on it has been preserved, but the linkskipped version hasn't

Does linkskipping still work with net7.0-ios projects, or is there a different way that its supposed to be done now?

Attached is the project I used for this

[testlinker.zip](https://github.com/xamarin/xamarin-macios/files/12353544/testlinker.zip)

**Comment by rolfbjarne** [comment]
In .NET, you use add the assembly to the list of rooted assemblies like this: https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-7-0#root-assemblies

```xml
<ItemGroup>
  <TrimmerRootAssembly Include="MyAssembly" />
</ItemGroup>
```

can you try that and see if it works for you?

**Comment by dlprows** [comment]
That seems to be working for top level references that would previously be linkskipped. However, I'm still running into issues.

Its my understanding that the way linkskipping used to work, if you linkskipped a top level reference, it would implicitly linkskip its dependencies.

With the TrimmerRootAssembly it looks like it doesn't do that.

In our large project that we're migrating to dotnet, even after converting the list of linkskipped assemblies with a list of TrimmerRootAssembly inclusions, I'm getting errors that look like things getting trimmed out.

Is this a difference in how mtouch linker worked vs how the trimmer works, or could i be running into something else?

**Comment by dlprows** [comment]
After putting together a different test project to validate my believe that linkskipping will implicitly linkskip dependencies, it looks like i'm wrong.

But there is still some difference in how this worked with mtouch, that i'm missing, because our project works fine with xamarin.ios with linking set to full, and the linkskips. But in net7 changing the linkskips to TrimmerRootAssembly it doesn't work

I'll see if i can build a sample project that does the same

**Comment by rolfbjarne** [comment]
> But there is still some difference in how this worked with mtouch, that i'm missing, because our project works fine with xamarin.ios with linking set to full, and the linkskips.

I don't know what you're running into, but one potential difference is that if you needed a type X from assembly A, and in .NET that type moved to assembly B, then the old linkskip for assembly A won't do what you need anymore.

> I'll see if i can build a sample project that does the same

That would be great! Marking as need-info until then.

**Comment by dlprows** [comment]
It looks like you're right and its not a difference in the thinning, but a difference in something being in a different assembly.
Using DotPeek, i found that the TrimmerRootAssembly really is preserving the classes that I thought were still getting trimmed.

Unfortunately I could never figure out what exactly it was that was moved. However, the reason we had the linker settings so aggressive was because our app was exceeding the amount of code in an app for app store submission. But it looks like they changed the limits years ago, and we just didn't keep up because the linking was working

https://developer.apple.com/help/app-store-connect/reference/maximum-build-file-sizes/

It looks like using SdkOnly for the link mode should be adequate for our use case, and if we run into issues with that, I got all the answers i needed for how to work with linking/thinning now.

Thank you so much!
